Sources/NIOPosix/PosixSingletons+ConcurrencyTakeOver.swift

```diff
@@ -0,0 +1,107 @@
+import Atomics
+import NIOCore
+
+private protocol SilenceWarning {
+    @available(macOS 14.0, iOS 17.0, watchOS 10.0, tvOS 17.0, *)
+    func enqueue(_ job: UnownedJob)
+}
+@available(macOS 14, *)
+extension SelectableEventLoop: SilenceWarning {}
+
+private let _haveWeTakenOverTheConcurrencyPool = ManagedAtomic(false)
+extension NIOSingletons {
+    /// Install ``MultiThreadedEventLoopGroup/singleton`` as Swift Concurrency's global executor.
+    ///
+    /// This allows to use Swift Concurrency and retain high-performance I/O alleviating the otherwise necessary thread switches between
+    /// Swift Concurrency's own global pool and a place (like an `EventLoop`) that allows to perform I/O
+    ///
+    /// This method uses an atomic compare and exchange to install the hook (and makes sure it's not already set). This unilateral atomic memory
+    /// operation doesn't guarantee anything because another piece of code could have done the same without using atomic operations. But we
+    /// do our best.
+    ///
+    /// - warning: You may only call this method from the main thread.
+    /// - warning: You may only call this method once.
+    @discardableResult
+    public static func unsafeTryInstallSingletonPosixEventLoopGroupAsConcurrencyGlobalExecutor() -> Bool {
```

Comment:
Thx for prefixing with unsafe

Reply:
Yes thought `unsafe` + a loooong name is okay, can easily make it disappear. Also it's a `*Try*` function. Nothing wrong to just turn this into a `return false` in a few years.

Reply:
What if we hide it behind an SPI? We have been recently discussing using an SPI for exactly those scenarios where we want users to make a very very conscious choice about it

Reply:
Hmm, this is not a complex API (just one function) so I don't think we need to use SPI for that. If we deprecate that, it doesn't leave a huge wrinkle on the NIO API surface.

Reply:
Should we document when not to do this?  Should we underscore the name?  Presumably it's a poor trade for most people unless they really know they want it.  I'm guessing likely disastrous for anyone running a UI app.

Reply:
I don't know if unsafe is strong enough. It fails to communicate to users that this relies on implementation details of the Swift runtime (`dlsym`ing symbols) and unsupported language features (`@convention(thin)`) as well as a bunch of assumptions about the layout of the runtime's non-public types.

It's not just that calling this might do weird junk, it's that calling it might explode in new Swift releases, and it would very much be our fault.

I'm inclined to say that we also want to underscore this, and expand on the doc comment to say that this function cannot guarantee that it will safely work on newer Swift compilers or any non-stable Swift release.

Reply:
What about leaving it `unsafe*` (no underscore) but making it only perform work on `#if compiler(>=5.9) && compiler(<5.10)`. Then we can gradually relax it as new Swift versions get released? And maybe at some point we can stop this continuous version-upping process once Swift implements enough to do this properly.

Reply:
FWIW, I don't like the compiler guards because it will continuously break and we need to play catch up with the compiler version. IMO `_` or `SPI` works for me here and indicates that this is really not something we want people to actively do.

Reply:
But nothing breaks here, it'll just return `false` and be slower. That's better than having `_badStuff` that people use anyway which then leads to crashes.

IMHO we can have either of two arguments:

1. Future Swift versions will likely break us here -> we should disable on future Swifts
2. Future Swift versions will likely not break us here -> `unsafe*`'s enough

I don't think offering useful functionality behind a `_` that may crash is ideal. If we're not sure if something will continue to work we should disable it for anything but known-good compilers, no? We'd then just force the `#if compiler(...)` into our users' codebases.

Reply:
I'm happy with that.

Reply:
@Lukasa implemented